### ISSUE TITLE
From Regex Not Working

### ISSUE THREAD

**Issue by x011** [RealRaven2000/FiltaQuilla#244]
I'm trying to match emails from the following senders (spammer):

```
info@s2.alexttr.ru
info@s6.alexspo.ru
info@alexspo.ru
info@s9.alexttr.ru
```

I've tested the following regex and it seems to match all cases:

`.*@.*alex.*?.ru`

But when I use it on Thunderbird Filter `Header Regex Match`, it doesn't work. 
Any Idea why?

![image](https://github.com/RealRaven2000/FiltaQuilla/assets/4313821/ba63abd9-7e20-47db-bed7-e4c927c9c8b7)

Headers snippet:

```
From: Aleksandr <info@s9.alexttr.ru>
Reply-To: Aleksandr <info@s9.alexttr.ru>
```

PS: I'm testing the filter by clicking on the "Run Now" button after selecting the desired folder:

![image](https://github.com/RealRaven2000/FiltaQuilla/assets/4313821/78342e09-33bf-4bbd-984e-94aad13d33c1)

**Comment by RealRaven2000** [comment]
There seems to be a problem with certain headers not appearing in the `nsIMsgFolder.getStringProperty()` anymore. I can try rewrite the code to find some of these missing properties in the `nsIMsgFolder.propertie`s array, but I have a feeling that a lot of them were removed - even a lot of the X- headers are missing at this stage. let's try and fix the "from" header first and take it from there...

**Comment by RealRaven2000** [comment]
Could you upload an example email from the spammer (just save as eml format and put into a zip file) so I can test please?

**Comment by RealRaven2000** [comment]
I think you need to use `sender` (lowercase) and not `from` - here is what is available:

```
Array(28) [ "flags", "sender", "recipients", "subject", "message-id", "dateReceived", "date", "priority", "size", "threadParent", … ]
​
0: "flags"
1: "sender"​
2: "recipients"​
3: "subject"​
4: "message-id"​
5: "dateReceived"​
6: "date"​
7: "priority"​
8: "size"​
9: "threadParent"​
10: "msgThreadId"​
11: "ProtoThreadFlags"
12: "X-GM-MSGID"​
13: "X-GM-THRID"​
14: "X-GM-LABELS"​
15: "msgOffset"​
16: "storeToken"​
17: "offlineMsgSize"​
18: "numLines"​
19: "junkscore"​
20: "junkscoreorigin"​
21: "junkpercent"​
22: "preview"​
23: "keywords"​
24: "gloda-id"​
25: "sender_name"​
26: "gloda-dirty"​
27: "remoteContentPolicy"​
length: 28
```
​

**Comment by RealRaven2000** [comment]
Here is a test version with more information in the console:

[filtaquilla-4.0pre40.zip](https://github.com/RealRaven2000/FiltaQuilla/files/13793158/filtaquilla-4.0pre40.zip)

to activate the debug log for header regex, open FiltaQuilla options and then open the About FQ/Help page.
Activate debug Mode (bottom right)
right-click debug mode and then activate extensions.filtaquilla.debug.regexHeader = true

![image](https://github.com/RealRaven2000/FiltaQuilla/assets/6314674/14997890-daf0-462b-be50-9bd0ad36c369)

![image](https://github.com/RealRaven2000/FiltaQuilla/assets/6314674/b09fe9ac-82a8-41da-8d02-1cd7ac47c5e3)

----

to install the test version, download the zip file and then drag the file into Thunderbird Add-ons Manager (do not extract contents).

**Comment by RealRaven2000** [comment]
did you get anywhere with testing? I would like to close old issues in order to focus on stuff that can be fixed.

**Comment by mat926** [comment]
I'm having a similar issue with Header Regex Match and also can't figure out why some of the filter doesn't work correctly.  Like the OP I am also creating filters for deleting spam emails. I have this following regex: 
`(?=(?:[^A-Z@]*[A-Z]){3,})(?=(?:[^a-z@]*[a-z]){3,})[A-Za-z]{6,10}@[^@]*\.[^@]*\.[^@]*` 
Which should match 3 lowercase & uppercase characters before the @ symbol , also the length before @ shouldn't be greater than 10 . And also checks for 2 periods after the @.  
For example it should match something like this: `cKefJUxC@space.smartsdirect.us` . 
But it also incorrectly matches this: `noreply@recruiting.facebook.com` (no uppercase characters) 
and this: `Yzgixpe@space.smartsdirect.us` 
 (Only 1 uppercase character before `@`. Please see the screenshot:

<img width="2502" height="1281" alt="Image" src="https://github.com/user-attachments/assets/203c0901-f427-4bb4-8783-1b49ff593d52" />

**Comment by RealRaven2000** [comment]
If you’re counting uppercase/lowercase letters using lookaheads, you definitely need to **disable case-insensitive matching**. Try:

`/(?=(?:[^A-Z@]*[A-Z]){3,})(?=(?:[^a-z@]*[a-z]){3,})[A-Za-z]{6,10}@[^@]*\.[^@]*\.[^@]*/c`

(`/c` is FiltaQuilla's pseudo tag for doing case sensitive matching)

you should also try and disable the “Regex case-insensitive match” option in FiltaQuilla settings (I think it appends `/i `if you do not declare your own `/options` at the end of your regex) :

<img width="761" height="175" alt="Image" src="https://github.com/user-attachments/assets/acedb833-4567-4eae-a140-8becdf85ff0e" />

----

if you want me to do a live test - can you share the eml with the false positives (save as eml file) and export the filter using quickFilters? 

<img width="617" height="192" alt="Image" src="https://github.com/user-attachments/assets/00c4979d-35e5-47ac-af15-e54109507a7d" />

it looks like it should match according to regex101 - which also uses JS (ECMAScript) flavor:

<img width="1052" height="207" alt="Image" src="https://github.com/user-attachments/assets/74b6dad2-8cd7-43b1-a74d-3a5fea0d8e2e" />

Screenshots of logs aren't super useful - they are hard to read an analyse - better to right-click console and select "copy all messages". at least this will copy the complete text.

**Comment by mat926** [comment]
I tried with setting the `/c` flag and I get the error in the console: 
```
Uncaught SyntaxError: invalid regular expression flag c
    headerRegEx_match chrome://filtaquilla/content/filtaquilla.js:1891
    runSelectedFilters chrome://messenger/content/FilterListDialog.js:741
    oncommand chrome://messenger/content/FilterListDialog.xhtml:1
```

The workaround for me is to disable Regex case-insensitive Match like you said , but it's a little inconvenient because I still want to enable insensitive match apply globally but only disable it for this specific filter. 

For context , here's the full expression I'm using: `sender:/(?=(?:[^A-Z@]*[A-Z]){3,})(?=(?:[^a-z@]*[a-z]){3,})[A-Za-z]{6,10}@[^@]*\.[^@]*\.[^@]*/c`

<img width="1545" height="777" alt="Image" src="https://github.com/user-attachments/assets/eab6ef1c-83d1-4d84-9b64-d202341c0f01" />

I've attached some emails and the filter so you can run some tests. 

[filtaquilla.zip](https://github.com/user-attachments/files/25189649/filtaquilla.zip)

**Comment by RealRaven2000** [comment]
> I tried with setting the `/c` flag and I get the error in the console:

sorry I only saw it in code, didn't actually test it - will report back:

```js
  getRegex: function (aSearchValue) {
    const regexpCaseInsensitiveEnabled = this.prefs.getBoolPref("regexpCaseInsensitive.enabled");
    const REGEX_CASE_SENSITIVE_FLAG = "c";
    let searchValue = aSearchValue,
      searchFlags = "",
      searchOptions = [];
    if (aSearchValue.charAt(0) == "/") {
      let lastSlashIndex = aSearchValue.lastIndexOf("/");
      if (!lastSlashIndex) {
        console.log(`invalid regex - missing second '/' ! ${aSearchValue}`);
        searchValue = aSearchValue.substring(1);
      } else {
        searchFlags = aSearchValue.substring(lastSlashIndex + 1);
        searchValue = aSearchValue.substring(1, lastSlashIndex);
      }
      let sw = searchFlags.match(/{.*}/) || [];
      if (sw && sw.length) {
        const startOptions = searchFlags.indexOf(sw[0]),
          optionString = searchFlags.substring(startOptions + 1, startOptions + sw[0].length - 1);
        searchOptions = optionString.split(",");

        searchFlags = searchFlags.substring(0, startOptions);
      }
    }

    if (
      regexpCaseInsensitiveEnabled &&
      !searchFlags.includes("i") &&
      !searchFlags.includes(REGEX_CASE_SENSITIVE_FLAG)
    ) {
      searchFlags += "i";
    }

    return [searchValue, searchFlags, searchOptions];
  },
```

**Comment by RealRaven2000** [comment]
Here is a patched version that removes the tag - also prints out more detail during header matching in console:

[filtaquilla-6.2pre17.zip](https://github.com/user-attachments/files/25209166/filtaquilla-6.2pre17.zip)

You need to restart Thunderbird so that the filter service loads the new search method. This gives me the following results:

<img width="691" height="253" alt="Image" src="https://github.com/user-attachments/assets/409edede-a8e5-4f99-801e-57471bf6daae" />

does this match your expectations for the regex?

---

To test the version above, download the zip file, drag the file into Thunderbird Add-ons Manager, do not extract contents or if won't install.

**Comment by mat926** [comment]
> does this match your expectations for the regex?

Yes i can confirm the case sensitive matching works as expected , thank you.
